### PR TITLE
WT-10831 Improve cache tracking for small uncompressed pages

### PR DESCRIPTION
**Pull request by Siddhartha8899**
Create a new variable (`dsk_alloc_size`) in the `WT_PAGE` structure. This variable is solely used for cache tracking purposes. 

Earlier we would use `WT_PAGE::dsk->mem_size` for cache tracking, however, this variable contains the actual number of bytes consumed by the disk image, not the size that was truly allocated.

The reason for the above was that we did not have the allocated memory value when discarding the page image. However, now we do (`dsk_alloc_size`).

### CODE REVIEW

File: src/btree/bt_page.c [hunk 1]
```diff
@@ -416,7 +416,7 @@ __wt_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, uint32
      * used to instantiate the page image even though the values might not match exactly, because
      * that's the only value we have when discarding the page image and accounting needs to match.
      */
-    size = LF_ISSET(WT_PAGE_DISK_ALLOC) ? dsk->mem_size : 0;
+    size = LF_ISSET(WT_PAGE_DISK_ALLOC) ? page->dsk_alloc_size = image_alloc_size : 0;
```
Comment (agorrod) [inline src/btree/bt_page.c:419]: Is this right? It will result in `size = page->dsk_alloc_size = image_alloc_size;` It's weird to do that in a ternary (cascading assignments is something I need to think about anyway, putting them in a ternary makes me wonder if it is deliberate).

File: src/btree/bt_slvg.c [hunk 2]
```diff
@@ -689,7 +689,7 @@ __slvg_trk_leaf(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, uint8_t *ad
          * Page flags are 0 because we aren't releasing the memory used to read the page into memory
          * and we don't want page discard to free it.
          */
-        WT_ERR(__wt_page_inmem(session, NULL, dsk, 0, &page, NULL));
+        WT_ERR(__wt_page_inmem(session, NULL, dsk, 0, 0, &page, NULL));
```
Comment (agorrod) [inline src/btree/bt_slvg.c:692]: I think passing `0` here will mean that the code goes from tracking the read size to tracking `0` when called via salvage. I expect that would make salvage cache tracking wrong.

File: src/btree/bt_read.c [hunk 3]
```diff
@@ -194,7 +194,7 @@ __page_read(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags)
     page_flags = WT_DATA_IN_ITEM(&tmp) ? WT_PAGE_DISK_ALLOC : WT_PAGE_DISK_MAPPED;
     if (LF_ISSET(WT_READ_IGNORE_CACHE_SIZE))
         FLD_SET(page_flags, WT_PAGE_EVICT_NO_PROGRESS);
-    WT_ERR(__wt_page_inmem(session, ref, tmp.data, page_flags, &notused, &prepare));
+    WT_ERR(__wt_page_inmem(session, ref, tmp.data, tmp.size, page_flags, &notused, &prepare));
```
Comment (agorrod) [inline src/btree/bt_read.c:197]: Is this always the right size? At the moment, the correct size is used for pages that were compressed, and I haven't done the work to understand why, so I'm asking you because I'm lazy:
Will this be correct for cases where the page was block compressed and not block compressed?

**Review by agorrod — COMMENTED** [review]
Thanks Sid - this is moving in the direction we want.
I have some questions - which made me wonder if there is a mechanism we could use to confirm that tracking of disk image size in the cache is correct. i.e: could we add testing that ensures we are tracking the right size?

File: src/btree/bt_page.c (hunk 1 above)
Comment (Siddhartha8899) [reply src/btree/bt_page.c:419]: It appears correct to me. If the `WT_PAGE_DISK_ALLOC` flag is not set, we intend to maintain `page->dsk_alloc_size` as 0 without altering it. We would only modify `page->dsk_alloc_size` and track the `size` if `WT_PAGE_DISK_ALLOC` is set.

If readability is your concern I am happy to change it.

File: src/btree/bt_slvg.c (hunk 2 above)
Comment (Siddhartha8899) [reply src/btree/bt_slvg.c:692]: Since the `WT_PAGE_DISK_ALLOC` flag is not set here, we would not track the size regardless.

File: src/btree/bt_read.c (hunk 3 above)
Comment (Siddhartha8899) [reply src/btree/bt_read.c:197]: Yes, I read the code and the size seemed correct. I will spend some more time on this.

File: src/btree/bt_split.c [hunk 4]
```diff
@@ -1395,7 +1395,8 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
      * our caller will not discard the disk image when discarding the original page, and our caller
      * will discard the allocated page on error, when discarding the allocated WT_REF.
      */
-    WT_RET(__wt_page_inmem(session, ref, multi->disk_image, WT_PAGE_DISK_ALLOC, &page, &prepare));
+    WT_RET(__wt_page_inmem(
+      session, ref, multi->disk_image, multi->size, WT_PAGE_DISK_ALLOC, &page, &prepare));
```
Comment (Siddhartha8899) [inline src/btree/bt_split.c:1399]: I think `multi->size` is the right variable to use here. However, a confirmation would be nice.

**Comment by Siddhartha8899** [comment]
@agorrod I am still thinking of a test to verify the cache tracking. I have run tests and debugged the code to ensure that I see the correct size.

**Comment by Siddhartha8899** [comment]
I have added a small python test for testing whether the cache tracks the page images correctly.

File: src/btree/bt_read.c (hunk 3 above)
Comment (keitharnoldsmith) [reply src/btree/bt_read.c:197]: To answer @agorrod's question:

> the correct size is used for pages that were compressed, and I haven't done the work to understand why, so I'm asking you because I'm lazy

When we read a compressed page, we read it into a buffer that matches the on-disk size. Then we read the in-memory size from the page header (which isn't compressed). We allocate a new buffer using this size and decompress the page into that new buffer. So the we wind up with a buffer of the correct size.

File: src/include/btmem.h [hunk 5]
```diff
@@ -716,17 +716,23 @@ struct __wt_page {
...
+    const WT_PAGE_HEADER *dsk;
+
     /*
      * !!!
      * This is the 64 byte boundary, try to keep hot fields above here.
      */
 
+    /*
+     * The allocated memory for the page-disk image, solely used for cache tracking purposes. The
+     * 'dsk' variable above contains the actual number of bytes consumed by the disk image.
```
Comment (keitharnoldsmith) [inline src/include/btmem.h:732]: I find the terminology confusing/ambiguous here. In particular, on my first reading, I interpreted "disk image" as the thing we read from disk -- which is the larger container for the actual bytes we use to construct the in memory page.

Maybe reword this as:
```
The allocated memory for the page's disk image, solely used for cache tracking purposes. The 'dsk' variable
above contains the actual size of the page contained in the disk image.
```

File: src/btree/bt_split.c (hunk 4 above)
Comment (keitharnoldsmith) [reply src/btree/bt_split.c]: I could dig into this and form an opinion, but it might be easier to pull in someone who is already familiar with the split code here.

File: src/include/btmem.h (hunk 5 above)
Comment (Siddhartha8899) [reply src/include/btmem.h]: Thanks, updated in [50b19cf](https://github.com/wiredtiger/wiredtiger/pull/9279/commits/50b19cf7349c635f8aaed022f121a76cde3b6680)

File: src/btree/bt_split.c (hunk 4 above)
Comment (Siddhartha8899) [reply src/btree/bt_split.c]: After discussing with @sulabhM, we found that the size allocated for creating a disk image for the multi structure is stored in multi->disk_image->mem_size (in __wt_rec_split). Therefore, I will utilize the same for cache tracking. 

Essentially, we do not plan to alter the cache tracking behavior when pages are created from a multi (disk image).

File: src/btree/bt_page.c [hunk 6]
```diff
@@ -329,8 +329,8 @@ __wt_page_inmem_prepare(WT_SESSION_IMPL *session, WT_REF *ref)
  *     Build in-memory page information.
  */
 int
-__wt_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, uint32_t flags,
-  WT_PAGE **pagep, bool *preparedp)
+__wt_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, size_t image_alloc_size,
+  uint32_t flags, WT_PAGE **pagep, bool *preparedp)
```
Comment (agorrod) [inline src/btree/bt_page.c:333]: I've been thinking about this change, because I'm concerned that it isn't obvious when we decide which size to track. Is it possible to make the change without this associated change to the input parameters here? i.e: call `wt_page_inmem` figure out if it should use the image allocation size or the disk allocation size?

Or, maybe there could be a variable in the WT_PAGE structure that always has the tracked size, and this function just uses that?

Comment (Siddhartha8899) [reply src/btree/bt_page.c:333]: Well, the key point is to consistently track the malloced size. In this function, the page is created within the function itself, meaning it hasn't been created yet. 

Determining the `image_alloc_size` within this function is a challenge if it is not directly passed as a parameter. However, I will spend some time thinking if there is a way.

Comment (Siddhartha8899) [reply src/btree/bt_page.c:333]: It appears challenging to incorporate this parameter into WT_REF (as we aim to avoid increasing the size of this struct). Another concern lies with the void* image, (which is persisted on disk, and modifying it may lead to compatibility issues). Additionally, the WT_PAGE cannot be modified since it has not been created yet.

One potential solution is to pass the WT_ITEM variable (instead of the const void*) and extract the image and size from it. However, this approach could pose problems if the function is called from `__split_multi_inmem`, as it does not have a WT_ITEM struct and instead uses WT_MULTI.

Comment (Siddhartha8899) [reply src/btree/bt_page.c:333]: @agorrod after re-reviewing this, I think it is best to pass the parameter (explicitly) itself.

File: test/suite/test_cache_tracking01.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import wiredtiger, wttest
+from wtscenario import make_scenarios
+from wtdataset import SimpleDataSet
+
+# Ensure cache is tracked correctly for page disk images.
```
Comment (keitharnoldsmith) [inline test/suite/test_cache_tracking01.py:33]: I would expand on this to be more specific about what this test is doing. Maybe something like:

`Ensure that cache size is tracked correctly when the content on a disk page is smaller than the page itself`

File: test/suite/test_cache_tracking01.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        # Read a page back in memory.
+        cursor.set_key(ds.key(1))
+        cursor.search()
+
+        cache_bytes_page_image = self.get_stat(wiredtiger.stat.conn.cache_bytes_image)
+
+        if self.compressor == "none":
+            self.assertGreaterEqual(cache_bytes_page_image, _128MB)
+        else:
+            self.assertLess(cache_bytes_page_image, _128MB)
```
Comment (keitharnoldsmith) [inline test/suite/test_cache_tracking01.py:89]: I would also add a comment here explaining the logic here. It is easy to see what comparisons the test makes. But for someone not familiar with the issue, it probably won't be clear why the test asserts these things.

I.e., explain that when there is compression, we allocate pages the amount of memory needed for the in-memory page, but when we have a page that isn't compressed, we allocate space based on the on-disk size rather than the in-memory size. This test is set up, so that 

Also, I would expect that the `assertGreaterEqual` here could just be `assertGreater`. To read the key from disk, we'll load both the root page and the leaf page, so we should have at least two 128MB pages in memory.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM, with some suggestions for additional comments in the Python test.

**Review by agorrod — COMMENTED** [review]
Sorry about the extended delay @Siddhartha8899 - I'm happy with this change. I've taken a run at different structures (because I wanted this to be more obvious) and couldn't find anything better. Well done!

**Review by agorrod — APPROVED** [review]

File: test/suite/test_cache_tracking01.py (hunk 8 above)
Comment (Siddhartha8899) [reply test/suite/test_cache_tracking01.py]: Updated in [9ace2ef](https://github.com/wiredtiger/wiredtiger/pull/9279/commits/9ace2effd9697c5bcf503a644f5829d49c8f58c2)
